src/dmd/mtype.d

```diff
@@ -5752,16 +5748,90 @@ extern (C++) final class TypeReturn : TypeQualified
 
 /***********************************************************
  */
-extern (C++) final class TypeStruct : Type
+extern (C++) abstract class TypeAggregate : Type
 {
-    StructDeclaration sym;
+    AggregateDeclaration asym;
+    alias sym = asym;
+
     AliasThisRec att = AliasThisRec.fwdref;
+
+    extern (D) this(TY ty, AggregateDeclaration asym)
+    {
+        assert(ty == Tstruct || ty == Tclass);
+
+        super(ty);
+        this.asym = asym;
+    }
+
+    override Dsymbol toDsymbol(Scope* sc)
+    {
+        return asym;
+    }
+
+    extern (D) abstract MATCH implicitConvToWithoutAliasThis(Type to);
```

Comment:
Just merge the implementations to keep it `final`.

You can put them as nested functions and switch on `ty` to call one or the other.

Reply:
Does final have a worse performance due to vtables?

Reply:
final is inlinable, unlike virtual (without some extra optimization pass, such as LTO).  But mostly it means that these extern(D) methods must be present in the C++ interface headers to occupy a slot in the vtable, but are not callable.